workers/unity/Packages/com.improbable.gdk.playerlifecycle/Systems/PlayerHeartbeat/HandlePlayerHeartbeatResponseSystem.cs

```diff
@@ -4,87 +4,78 @@
+            public EntityArray Entities;
+            [ReadOnly] public ComponentDataArray<PlayerHeartbeatClient.Component> NeedsHeartbeat;
+            public SubtractiveComponent<HeartbeatData> NoHeartbeatData;
+        }
+
+        private struct HeartbeatGroup
+        {
+            public readonly int Length;
+            public EntityArray Entities;
+            public ComponentDataArray<HeartbeatData> Heartbeats;
             public ComponentDataArray<WorldCommands.DeleteEntity.CommandSender> WorldCommandSenders;
+            [ReadOnly] public ComponentDataArray<PlayerHeartbeatClient.CommandResponses.PlayerHeartbeat> PlayerHeartbeatResponses;
             [ReadOnly] public ComponentDataArray<SpatialEntityId> SpatialEntityIds;
-            [ReadOnly] public ComponentDataArray<AwaitingHeartbeatResponseTag> AwaitingHeartbeatResponses;
-            public EntityArray Entities;
         }
 
-        [Inject] private Data data;
+        [Inject] private InitializeGroup initializeGroup;
+        [Inject] private HeartbeatGroup heartbeatGroup;
 
         protected override void OnUpdate()
         {
-            for (var i = 0; i < data.Length; i++)
+            for (var i = 0; i < initializeGroup.Length; i++)
+            {
+                PostUpdateCommands.AddComponent(initializeGroup.Entities[i], new HeartbeatData());
+            }
+
+            for (var i = 0; i < heartbeatGroup.Length; i++)
             {
-                var entityId = data.SpatialEntityIds[i].EntityId;
-                var responses = data.PlayerHeartbeatResponses[i].Responses;
-                var entity = data.Entities[i];
-                var entityDeleteSender = data.WorldCommandSenders[i];
+                var heartbeatsComponent = heartbeatGroup.Heartbeats[i];
 
+                var stillAlive = false;
+                var responses = heartbeatGroup.PlayerHeartbeatResponses[i].Responses;
                 foreach (var response in responses)
                 {
-                    if (response.StatusCode != StatusCode.Success)
+                    if (response.StatusCode == StatusCode.Success)
                     {
-                        var failedHeartbeatsComponent = EntityManager.HasComponent<HeartbeatData>(entity)
-                            ? EntityManager.GetComponentData<HeartbeatData>(entity)
-                            : new HeartbeatData();
-                        failedHeartbeatsComponent.NumFailedHeartbeats++;
-                        if (EntityManager.HasComponent<HeartbeatData>(entity))
-                        {
-                            PostUpdateCommands.SetComponent(entity, failedHeartbeatsComponent);
-                        }
-                        else
-                        {
-                            PostUpdateCommands.AddComponent(entity, failedHeartbeatsComponent);
-                        }
+                        stillAlive = true;
+                        break;
```

Comment:
Is there a case where `break` drops information from a subsequent command responses?

Reply:
Nop. Most of the time this system will run with just 1 response per entity, because it only sends them every X seconds. In the weird case more then 1 comes in, we only care about if all of the failed, to avoid false positives.